thijsvanloef/palworld-server-docker
# Add Discord player logging to notify when players join and leave the server

## Context 

Add discord player logging to notify via Discord webhook when players join and leave the server.

Adds this feature mentioned in #324 

> Script for notifying the public join/leave channel and broadcasting via RCON when a player joins or leaves: ![image](https://private-user-images.githubusercontent.com/41028418/304623480-b1ace498-bd68-4e95-8c66-4273597bed9f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDk1MDE1MjQsIm5iZiI6MTcwOTUwMTIyNCwicGF0aCI6Ii80MTAyODQxOC8zMDQ2MjM0ODAtYjFhY2U0OTgtYmQ2OC00ZTk1LThjNjYtNDI3MzU5N2JlZDlmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAzMDMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMzAzVDIxMjcwNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTMyNTE5MTUxY2Q1MDE5YjQ0YTk5NDFmMTJjNWQ1ZWE3NWQyNzc3MDRmNWZjYzcxMWZhN2VlNWQ5ZDEyMDg4ZTImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.sjjcrQGss6O2FTlI2a_0AHJjZvYzYcuxV5Xg7EzD930)


## Choices

I only 1 line for joining and one line for leaving inside the pre-existing player_logging.sh loops that are currently used for broadcasting to the server. I added lines to call the pre-existing function DiscordMessage() to send a message in the loop.

I added two env variables to allow the ability to change the join and leave messages. (DISCORD_PLAYER_LEAVE_MESSAGE and DISCORD_PLAYER_JOIN_MESSAGE)

I'm quite new to docker so I could be wrong, but I believe the way I did it would work. In the dockerfile I left ${player_name} in both messages for them to be substituted later. If I understand correctly, the string "${player_name}" can not be removed if the message is changed, otherwise the substitution would fail.

I did not add anything in the message to indicate how long a player was on the server before they left.

## Test instructions

I am unable to test it myself, therefore I haven't.

1.) Create a DISCORD_WEBHOOK_URL
2.) Have a player join or leave the server
3.) Check for a message in the channel for which the DISCORD_WEBHOOK_URL is set up to post to

## Checklist before requesting a review
